eodag/plugins/download/http.py

```diff
@@ -979,49 +946,50 @@ def _stream_download(
-                    self.stream, "status_code", None
-                ) is not None and self.stream.status_code == getattr(
-                    self.config, "order_status", {}
-                ).get(
-                    "ordered", {}
-                ).get(
-                    "http_code"
-                ):
-                    product.properties["storageStatus"] = "ORDERED"
-                    self._process_exception(None, product, ordered_message)
-                stream_size = self._check_stream_size(product) or None
-
-                product.headers = self.stream.headers
-                filename = self._check_product_filename(product) or None
-                product.headers[
-                    "content-disposition"
-                ] = f"attachment; filename={filename}"
-                content_type = product.headers.get("Content-Type")
-                guessed_content_type = (
-                    guess_file_type(filename) if filename and not content_type else None
-                )
-                if guessed_content_type is not None:
-                    product.headers["Content-Type"] = guessed_content_type
+        )
+        try:
+            self.stream.raise_for_status()
+        except requests.exceptions.Timeout as exc:
+            raise TimeOutError(exc, timeout=DEFAULT_STREAM_REQUESTS_TIMEOUT) from exc
+        except RequestException as e:
+            self._process_exception(e, product, ordered_message)
+            return None
+        else:
+            # check if product was ordered
+
+            if getattr(
+                self.stream, "status_code", None
+            ) is not None and self.stream.status_code == getattr(
+                self.config, "order_status", {}
+            ).get(
+                "ordered", {}
+            ).get(
+                "http_code"
+            ):
+                product.properties["storageStatus"] = "ORDERED"
+                self._process_exception(None, product, ordered_message)
+            stream_size = self._check_stream_size(product) or None
+
+            product.headers = self.stream.headers
+            filename = self._check_product_filename(product)
+            product.headers["content-disposition"] = f"attachment; filename={filename}"
+            content_type = product.headers.get("Content-Type")
+            guessed_content_type = (
+                guess_file_type(filename) if filename and not content_type else None
+            )
+            if guessed_content_type is not None:
+                product.headers["Content-Type"] = guessed_content_type
 
-                progress_callback.reset(total=stream_size)
+            progress_callback.reset(total=stream_size)
+
+            def iteration_wrapper():
                 for chunk in self.stream.iter_content(chunk_size=64 * 1024):
                     if chunk:
                         progress_callback(len(chunk))
                         yield chunk
 
+            return filename, iteration_wrapper
```

Comment:
instead of changing method signature, inspire from what was done in `_stream_download_assets` to store filename

Reply:
it might be a good idea to store information in a new `EOProduct.file_metadata` (being a `TypedDict` ?). 

Use the same thing for `Asset` and move `size`, `filename`, `rel_path` inside it

Reply:
From what I can see, within the context of `_stream_download` the product does not yet have any assets (i.e.  `product.assets` is an empty dictionary). So there is no properties attribute to which we can add the filename.

Reply:
to make things more simple, we can simply store the needed file information as new `EOProduct` attributes, and delay the usage of `file:` STAC extension later

Reply:
Addressed in commit https://github.com/CS-SI/eodag/pull/1400/commits/d6eae625a6b1bfb74a33d7ea0f182a3bd58506e8